src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,286 @@
+  hypers->beta = bayesmix::to_eigen(hyperscast.beta());
+  hypers->phi = hyperscast.phi();
+  hypers->q = hyperscast.q();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+MFAHierarchy::get_hypers_proto() const {
+  bayesmix::MFAPriorDistribution hypers_;
+  bayesmix::to_proto(hypers->mutilde, hypers_.mutable_mutilde());
+  bayesmix::to_proto(hypers->beta, hypers_.mutable_beta());
+  hypers_.set_alpha0(hypers->alpha0);
+  hypers_.set_phi(hypers->phi);
+  hypers_.set_q(hypers->q);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+  out->mutable_mfa_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void MFAHierarchy::sample_full_cond(bool update_params) {
+  if (this->card == 0) {
+    // No posterior update possible
+    sample_prior();
+  } else {
+    sample_eta();
+    sample_mu();
+    sample_psi();
+    sample_lambda();
+  }
+}
+
+void MFAHierarchy::sample_eta() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  auto sigma_eta_inv_llt =
+      (Eigen::MatrixXd::Identity(hypers->q, hypers->q) +
+       state.lambda.transpose() * state.psi_inverse * state.lambda)
+          .llt();
+  Eigen::MatrixXd sigma_eta(sigma_eta_inv_llt.solve(
+      Eigen::MatrixXd::Identity(hypers->q, hypers->q)));
+
+  if (state.eta.rows() != card) {
+    state.eta.resize(card, state.eta.cols());
+    state.eta = Eigen::MatrixXd::Zero(card, state.eta.cols());
+  }
+  Eigen::MatrixXd temp_product(sigma_eta * (state.lambda.transpose()) *
+                               state.psi_inverse);
+  auto iterator = cluster_data_idx.begin();
+  for (size_t i = 0; i < card; i++, iterator++) {
+    Eigen::VectorXd tempvector((*dataset_ptr).row(*iterator));
+    state.eta.row(i) = (bayesmix::multi_normal_prec_chol_rng(
+        temp_product * (tempvector - state.mu), sigma_eta_inv_llt.matrixL(),
+        rng));
+  }
+}
+
+void MFAHierarchy::sample_mu() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  Eigen::DiagonalMatrix<double, Eigen::Dynamic> sigma_mu;
+
+  sigma_mu.diagonal() =
+      (card * state.psi_inverse.diagonal().array() + hypers->phi)
+          .cwiseInverse();
+
+  Eigen::VectorXd sum = Eigen::VectorXd::Zero(dim);
+  for (size_t i = 0; i < card; i++) {
+    Eigen::VectorXd row = state.eta.row(i);
+    sum += state.lambda * row;
+  }
+  Eigen::VectorXd mumean = sigma_mu * (hypers->phi * hypers->mutilde +
+                                       state.psi_inverse * (data_sum - sum));
+  state.mu = bayesmix::multi_normal_diag_rng(mumean, sigma_mu, rng);
+}
+
+void MFAHierarchy::sample_lambda() {
+  auto& rng = bayesmix::Rng::Instance().get();
+
+  Eigen::MatrixXd temp_etateta(state.eta.transpose() * state.eta);
+
+  for (size_t j = 0; j < dim; j++) {
+    auto sigma_lambda_inv_llt =
+        (Eigen::MatrixXd::Identity(hypers->q, hypers->q) +
+         temp_etateta / state.psi[j])
+            .llt();
+    Eigen::VectorXd tempsum(card);
+    auto iterator = cluster_data_idx.begin();
```

Comment:
We are not creating a temporary matrix anymore to access the dataset by columns but we iterate through the `cluster_data_idx` `std::set`

Reply:
This sum looks weird to me. First iterator enters only in `dataset_ptr`, so you can divide by `state.psi[j]` only at the end saving `card` divisions.
Also subtracting `state.mu[j]` can be done only once.

I think a faster way to get what you want is
```
const Eigen::VectrorXd& data_col = dataset_ptr->col(j);
for (size_t i = 0; i < card; i++) {
       tempsum[i] = data_col(*iterator);
      iterator++;
}
tempsum -= state.mu[j];
temspum /= state.psi[j];
```


 "Array of indices" section here https://eigen.tuxfamily.org/dox/group__TutorialSlicingIndexing.html

Reply:
Also I would be consistent with the way you increase the iterator, usually you do it in the for loop declaration

Reply:
Finally, please add a //TODO comment, once the library is upgraded to Eigen 3.4.0, this will be much faster thanks to slicing

Reply:
The first implementation we had was with the new Eigen feature with slicing, but then we noticed it was not yet implemented in 3.3.9. But I think a `[]` operator is needed for custom slicing, so we had a `std::vector` as a copy of the `std::set` at each iteration. Do you think we have any other way around that?

Reply:
> This sum looks weird to me. First iterator enters only in `dataset_ptr`, so you can divide by `state.psi[j]` only at the end saving `card` divisions. Also subtracting `state.mu[j]` can be done only once.
> 
> I think a faster way to get what you want is
> 
> ```
> const Eigen::VectrorXd& data_col = dataset_ptr->col(j);
> for (size_t i = 0; i < card; i++) {
>        tempsum[i] = data_col(*iterator);
>       iterator++;
> }
> tempsum -= state.mu[j];
> temspum /= state.psi[j];
> ```
> 
> "Array of indices" section here https://eigen.tuxfamily.org/dox/group__TutorialSlicingIndexing.html

Yes, it should be faster the way you suggested if we do `tempsum -= state.mu[j];` with the coefficient-wise functions of Eigen instead of doing it explicitly in the for loop. The name tempsum is not appropriate since we are not actually summing in the for loop but `iterator` and `i` are just the same iterator for two different data structures. We have a few other places where this could be done.